### ISSUE TITLE
No reconnection after 'Connection reset by peer'

### ISSUE THREAD

**Issue by HankAviator** [shadowsocks/shadowsocks-android#1414]
Please read FAQ then answer these questions before submitting your issue. Thanks!

### Environment

* Android version: 7.1.2, LineageOS 14.1
* Device: OnePlus One
* Shadowsocks version: 4.2.5
* Last version that did not exhibit the issue (if applicable): N/A

### Configuration

_Put an `x` inside the [ ] that applies._

* [x] IPv4 server address
* [ ] IPv6 server address
* [x] Client IPv4 availability
* [ ] Client IPv6 availability
* Local port: 1080
* Encrypt method: RC4-MD5
* [ ] One-time authentication
* Route
  * [ ] All
  * [ ] Bypass LAN
  * [ ] Bypass China
  * [x] Bypass LAN & China
  * [ ] GFW List
  * [ ] China List
  * [ ] Custom rules
* [ ] IPv6 route
* [ ] Per-App Proxy
  * [ ] Bypass mode
* Remote DNS: 8.8.8.8
* [ ] DNS Forwarding
* Plugin configuration (if applicable):
* [x] Auto Connect
* [ ] TCP Fast Open
* [x] NAT mode

### What did you do?
Launch shadowsocks normally and use other apps. In a irregular phase, shadowsocks writes a line of
```
E shadowsocks: remote_recv_cb_recv: Connection reset by peer
```
into logcat. Right after this line, all network traffic is unavailable, including China, LAN and GFWed traffic.

*Note: On marshmallow, shadowsocks seems to be **correctly** resuming the connection*, though still sometimes it requires manual intervention (turn off then back on).
Tried to uninstall, reinstall (by offline apk via package manager or Google Play), clear data, checked task-kill utilities and all possible whitelists have ss included. Even set highest process priority and permanent for `shadowsocks:bg`.

### What did you expect to see?
Shadowsocks should try to reconnect even the connection is reset.

### What did you see instead?
No reconnection and all connections are obstructed.
No other lines in log are relevant so I'm not posting here. They are well kept and will be uploaded on demand.
App **IS NOT** killed, both`shadowsocks:bg` and `shadowsocks:sh` are still running.

**Comment by HankAviator** [comment]
shadowsocks has already been put into whitelist of Battery Optimization and Unlimited and Unrestricted data access (in data saver).

**Comment by madeye** [comment]
First of all, try VPN mode instead.

According to your description, it looks a network issue. To verify background processes are not killed, please run the following command in the adb shell:

`ps aux | grep shadowsocks`

**Comment by HankAviator** [comment]
VPN mode so far so good. Monitoring any further peculiar event.

**Comment by HankAviator** [comment]
NAT mode, log captured as below:
```
bacon:/ $ ps | grep shadowsocks
u0_a233   6494  361   1334824 71056 sys_epoll_ 00000000 S com.github.shadowsocks:bg
u0_a233   26391 6494  4600   1996  sys_epoll_ 00000000 S /data/app/com.github.shadowsocks-1/lib/arm/libredsocks.so
u0_a233   26395 6494  5504   2608  sys_epoll_ 00000000 S /data/app/com.github.shadowsocks-1/lib/arm/libss-local.so
u0_a233   26402 6494  843972 14768 futex_wait 00000000 S /data/app/com.github.shadowsocks-1/lib/arm/liboverture.so
```
process was alive.

**Comment by madeye** [comment]
NAT mode has been deprecated for years. Make sure you're using VPN mode at first.

**Comment by wongsyrone** [comment]
> using VPN mode

You might want to update tun2socks which have latest lwip included.

**Comment by HankAviator** [comment]
Update: VPN mode delayed some notifications (WeChat) while minutes later also stopping all traffic. This happened only after a night of screen off. Didn't get any log since it might oversize. Not sure if it's yielded by Doze. Turned off some experimental features in Greenify, reset Amplify restriction rules and I'll keep watching.

**Comment by HankAviator** [comment]
VPN Mode logcat when network unavailable (only interesting parts):
```
10-18 18:21:24.485  2065  2065 D tun2socks: ERROR(BConnection): connection failed
10-18 18:21:24.485  2065  2065 D tun2socks: ERROR(BSocksClient): connection failed
10-18 18:21:25.125  2065  2065 D tun2socks: ERROR(BConnection): connection failed
10-18 18:21:25.125  2065  2065 D tun2socks: ERROR(BSocksClient): connection failed
```
```
10-18 18:22:10.458  2065  2065 D tun2socks: ERROR(BConnection): recv failed
10-18 18:22:11.739  2065  2065 D tun2socks: ERROR(BConnection): recv failed
10-18 18:22:13.658  2065  2065 D tun2socks: ERROR(BConnection): recv failed
10-18 18:22:14.901  2184  2238 I ActivityManager: Displayed eu.chainfire.supersu/.PromptActivity: +374ms
10-18 18:22:14.967  1957  4258 I overture: time="2017-10-18T10:22:14Z" level=warning msg="Dial DNS upstream with SOCKS5 proxy failed: proxy: failed to read greeting from SOCKS5 proxy at 127.0.0.1:1080: read tcp 127.0.0.1:40174->127.0.0.1:1080: read: connection reset by peer" 
```

**Comment by madeye** [comment]
As per your log, the `ss-local` process should be killed already.

My suggestion is trying other ROMs, e.g. the official ROM from OnePlus.

**Comment by HankAviator** [comment]
I've run `ps` right after the network and the process list is the same as aforementioned in NAT mode debugging, processes are still there but seems the module within wouldn't function properly. Maybe Xposed framework caused some crash to socket yada yada as I see logcat was spammed by relevant errors.

**Comment by denniskrol** [comment]
Since I installed Xposed 88.1 I have the same problem. As soon as I disable Xposed (and reboot) Shadowsocks works again.

**Comment by HankAviator** [comment]
Reopened for reminder purpose.

**Comment by denniskrol** [comment]
This is kinda off-topic, but in case anybody ever finds this with Google:
Always Expandable Notifications (http://repo.xposed.info/module/com.mohammadag.alwaysexpandablenotifications) broke it. It fucks with the persistent notification, and Shadowsocks doesn't seem to like it.
All Notifications Expanded (http://repo.xposed.info/module/com.ugglynoodle.allnotificationsexpanded) does the same, and does work.
I enabled Amplify again (does that even work for Android 7?) and Shadowsocks still works.

**Comment by HankAviator** [comment]
In my case Amplify floods log and might due to unable to hook on some classes. Anyways, disabled for now. Monitoring other modules to see any can be also suspicious - or it was the framework itself.

**Comment by Mygod** [comment]
We are not going to support incompatibilities with Xposed framework and/or modules. Please report this bug to them instead.

Feel free to keep the conversation going in this thread.

**Comment by HankAviator** [comment]
It might be due to OOM on my old device OnePlus One (_bacon_) or LoS build too. Not sure which caused it.
On mi 6 (_sagit_) all is good even with xposed v89

**Comment by Sarun2515** [comment]
> เราจะไม่สนับสนุนความไม่เข้ากันกับกรอบงานและ/หรือโมดูล Xposed โปรดรายงานข้อบกพร่องนี้ให้พวกเขาทราบแทน
> 
> โปรดอย่าลังเลที่จะสนทนาต่อในกระทู้นี้

**Comment by Sarun2515** [comment]
### 

> **
Duplicate of #**
